src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineViewModel.cs

```diff
@@ -0,0 +1,443 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Collections;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Editor.Shared.Tagging;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Editor.Tagging;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.SemanticModelReuse;
+using Microsoft.CodeAnalysis.Shared.Collections;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.Internal.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.LanguageServer.Client;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Threading;
+using Roslyn.Utilities;
+using VsWebSite;
+
+namespace Microsoft.VisualStudio.LanguageServices.DocumentOutline
+{
+    using LspDocumentSymbol = DocumentSymbol;
+
+    /// <summary>
+    /// Responsible for updating data related to Document outline. It is expected that all public methods on this type
+    /// do not need to be on the UI thread. Two properties: <see cref="SortOption"/> and <see cref="SearchText"/> are
+    /// intended to be bound to a WPF view and should only be set from the UI thread.
+    /// </summary>
+    internal sealed partial class DocumentOutlineViewModel : INotifyPropertyChanged, IDisposable
+    {
+        private readonly ILanguageServiceBroker2 _languageServiceBroker;
+        private readonly ITaggerEventSource _taggerEventSource;
+        private readonly ITextView _textView;
+        private readonly ITextBuffer _textBuffer;
+        private readonly IThreadingContext _threadingContext;
+
+        /// <summary>
+        /// Queue that computes the new model and updates the UI state.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue _workQueue;
+
+        /// <summary>
+        /// Queue responsible for updating the ui after a change/move happens.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue _selectionQueue;
+
+        public event PropertyChangedEventHandler? PropertyChanged;
+
+        // Mutable state.  Should only update on UI thread.
```

Comment:
Why do we need this? can't we just have an immutable snapshot model?

Reply:
All teh models are immutable snapshots.  But this tracks moving the pointer to the new snapshot.  This is similar to ITextBuffer/ITextSnapshot.  The buffer points to the current (immutable) one.  but the buffer can be updated to a new one.

Reply:
for these pieces... they do appear to represent the current state of hte UI.  e.g. what current sort option is set (which needs to be stored somewhere), what the current contents of hte search-box are (which we need so we can filter).  What symbol-items we have computed and are presenting.